File: 015_MaximumNumberOfVowelsInSubstringOfGivenLength.py
"""
Topic: Sliding Window
Maximum Number of Vowels in Substring of Given Length: sliding window counting vowels
Time Complexity: O(n)
Space Complexity: O(1)
"""

def max_vowels(s: str, k: int) -> int:
    vowels = set('aeiou')
    cur = sum(1 for c in s[:k] if c in vowels)
    best = cur
    for i in range(k, len(s)):
        if s[i-k] in vowels: cur -= 1
        if s[i] in vowels: cur += 1
        best = max(best, cur)
    return best


if __name__ == '__main__':
    print(max_vowels('abciiidef', 3))  # 3
    print(max_vowels('aeiou', 2))  # 2

File: 015_MaximumNumberOfVowelsInSubstringOfGivenLength.cpp
// Topic: Sliding Window / String
// Maximum Number of Vowels in Substring of Given Length: Find the most vowels in any substring of
// length k. Time Complexity: O(n) Space Complexity: O(1)

#include <bits/stdc++.h>
using namespace std;

bool isVowel(char c)
{
    c = tolower(c);
    return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u';
}

int maxVowels(const string& s, int k)
{
    int n = s.size();
    int cnt = 0;
    for (int i = 0; i < min(n, k); ++i)
        if (isVowel(s[i]))
            ++cnt;
    int best = cnt;
    for (int i = k; i < n; ++i)
    {
        if (isVowel(s[i - k]))
            --cnt;
        if (isVowel(s[i]))
            ++cnt;
        best = max(best, cnt);
    }
    return best;
}

int main()
{
    cout << maxVowels("abciiidef", 3) << '\n';  // 3
    cout << maxVowels("aeiou", 2) << '\n';      // 2
    cout << maxVowels("leetcode", 3) << '\n';   // 2
    return 0;
}
